File: src/model/prescription.rb
#!/usr/bin/env ruby
# encoding: utf-8
# Prescriptions are not saved in the database to preserve the anonymity of our users.
# Therefor no mentioning of persistance and odba here
require 'date'
require 'model/package'
require 'model/doctor'
require 'util/today'

module ODDB
  class Prescription
    URL_FORMAT_DESCRIPTION  = 'http://2dmedication.org/'
    FORMAT_VERSION          = '1.0'
    SW_ORIGIN_ID            = 'ywesee GmBh'
    SW_VERSION_ID           = '1.0'
    QR_TIME_FORMAT          = '%Y%m%d'
    LOCAL_TIME_FORMAT       = '%d.%m.%Y'
    
    attr_reader :guid, # Eindeutiger Rezeptidentifier gemäss wikipedia.org/wiki/GUID
          :items 
    attr_accessor :doctor_glin,     # aka ean13
          :doctor_zsr,            # ZSR des ausstellenden Arztes
          :patient_id,            # Versichertenkartennummer des Patienten (VEKA)
          :date_issued,           # Datum der Rezeptausstellung
          :patient_family_name,
          :patient_first_name,
          :patient_zip_code,
          :patient_birthday,
          :patient_insurance_glin # aka  Patient KV EAN      
    
    class PrescriptionItem
      attr_accessor :ean13,
          :pharmacode,
          :description,
          :quantity,
          :valid_til,
          :simple_posology,
          :extended_posology,
          :nr_repetitions,
          :may_be_substituted
      def pretty
        "Rezept gültig bis: #{valid_til ? valid_til.strftime('%d.%m.%Y') : '' }\n" +
        "#{nr_repetitions}x [#{quantity}] #{ean13} #{pharmacode} #{description}"+
            "#{simple_posology ? "\n"+simple_posology : '' }#{extended_posology ? "\n"+extended_posology : '' }"
      end
      def is_valid?
        true
      end
      # initializes the item with sensible defaults
      def initialize
        @simple_posology = [0, 0, 0, 0]
        @may_be_substituted = true
        @quantity = 1
      end
    end
      
    # initializes the Prescription with sensible defaults
    def initialize
      @items = []
    end
    def add_item(prescription_item)
      raise "invalid prescription_item #{prescription_item}" unless prescription_item.is_valid?
      @items << prescription_item
    end
    def qr_string
      s = "#{URL_FORMAT_DESCRIPTION}|#{FORMAT_VERSION}|" +
          "#{guid}|#{SW_ORIGIN_ID}|#{SW_VERSION_ID}|" +
          "#{doctor_glin}|#{doctor_zsr}|#{patient_id}|#{date_issued.strftime(QR_TIME_FORMAT)}|#{patient_family_name}|#{patient_first_name}|"+
          "#{patient_zip_code}|#{patient_birthday.strftime(QR_TIME_FORMAT)}|#{patient_insurance_glin};"
      @items.each{ 
        |item|
      if item.ean13
        s +=  "#{item.ean13}|#{item.pharmacode}|#{ item.description}|"
      elsif item.pharmacode
        s +=  "#{item.ean13}|#{item.pharmacode}|#{ item.description}|"
      elsif item.ean13 == nil and item.description == nil 
        s +=  "#{item.ean13}|#{item.pharmacode}|#{ item.description}|"
      elsif item.description
        s +=  "||#{item.description}|"
      else
        raise "invalid item #{item.inspect}"
      end      
      s += item.quantity.to_s + '|'
      s += item.valid_til ? item.valid_til.strftime(QR_TIME_FORMAT)  + '|'    : '|'
      if item.simple_posology
        formatted =[]
        item.simple_posology.each{ |elem| formatted << sprintf('%0.2f', elem.to_f) }
        s += formatted.join('-') + '|'
      end
      s += item.extended_posology ? item.extended_posology + '|'              : '|'
      s += item.nr_repetitions ? item.nr_repetitions.to_s + '|'               : '|'
      s += item.may_be_substituted ? '0|' : '1|'
      }
      s = s.chomp('|')
      s += ';' + Prescription. checksum(s.clone).to_s
    end

    def pretty
      s = "Dr. #{doctor_glin} ZSR #{doctor_zsr}
Rezept ausgestellt am #{date_issued.strftime(LOCAL_TIME_FORMAT)}
Für #{patient_first_name} #{patient_family_name} ( #{patient_id})
Aus #{patient_zip_code} geboren am #{patient_birthday.strftime(LOCAL_TIME_FORMAT)}
Versicherungsnummer #{patient_insurance_glin}
"
      
      @items.each{ |item| s += item.pretty + "\n"  }
      s
    end

    def Prescription.checksum(s)
      # Fussdaten Prüfziffer Zahl (Long) Pflichtfeld ASCII Wertsumme des Textes, vor dem decodieren in Quoted Printable und ohne das letzte Feld und Trennzeichen
      checksum = 0
      0.upto(s.length-1).each{
        |idx|
          ascii = s[idx].ord 
          checksum += ascii
      }
      checksum
    end
  end
end

File: src/view/drugs/prescription.js
#!/usr/bin/env nodejs

module.exports = {
  
// procedure guid taken from http://stackoverflow.com/questions/105034/how-to-create-a-guid-uuid-in-javascript
/**
 * Generates a GUID string.
 * @returns {String} The generated GUID.
 * @example af8a8416-6e18-a307-bd9c-f2c947bbb3aa
 * @author Slavik Meltser (slavik@meltser.info).
 * @link http://slavik.meltser.info/?p=142
 */

guid: function() {
    function _p8(s) {
        var p = (Math.random().toString(16)+"000000000").substr(2,8);
        return s ? "-" + p.substr(0,4) + "-" + p.substr(4,4) : p ;
    }
    return _p8() + _p8(true) + _p8(true) + _p8();
},

PrescriptionItem: function (ean13) {
  this.ean13 = ean13;
  this.pharmacode = '';
  this.description = '';
  this.quantity = 1;
  this.valid_til = '';
  this.simple_posology = [0, 0, 0, 0];
  this.extended_posology = '';
  this.nr_repetitions = 0;
  this.may_be_substituted = true;
},

Prescription: function (guid) {
    this.URL_FORMAT_DESCRIPTION  = 'http://2dmedication.org/';
    this.FORMAT_VERSION          = '1.0';
    this.SW_ORIGIN_ID            = 'ywesee GmBh';
    this.SW_VERSION_ID           = '1.0';
    this.guid = guid;
    this.items = [];
    this.doctor_glin = '';           // aka ean13
    this.doctor_zsr = '';            // ZSR des ausstellenden Arztes
    this.patient_id = '';            // Versichertenkartennummer des Patienten (VEKA)
    this.date_issued = '';           // Datum der Rezeptausstellung
    this.patient_family_name = '';
    this.patient_first_name = '';
    this.patient_zip_code = '';
    this.patient_birthday = '';
    this.patient_insurance_glin
    this.to_s =  'A prescription is '+guid;
    
    this.qr_string = function() { 
      var s = this.URL_FORMAT_DESCRIPTION+'|'+this.FORMAT_VERSION+'|' +
      this.guid + '|' + this.SW_ORIGIN_ID+ '|'+this.SW_VERSION_ID+'|' +
      this.doctor_glin+'|' + this.doctor_zsr+'|' + this.patient_id+'|' + this.date_issued+'|' + 
      this.patient_family_name+'|' + this.patient_first_name+'|'+
      this.patient_zip_code+'|' + this.patient_birthday+'|' + this.patient_insurance_glin+';'
      for (var j = 0; j < this.items.length; j++) {
        var str = '';
        var item = this.items[j];
        if (item.ean13 != undefined) {
          str += item.ean13 + '|' + ( item.pharmacode != undefined ? item.pharmacode : '') + '|' + ( item.description != undefined ?  item.description : '') + '|';
        } else if (item.pharmacode != undefined) {
          str +=  '|' +  item.pharmacode + '|' + item.description + '|';
        } else if (item.description != undefined) {
          str +=  '||' + item.description +'|';
        } else  {
          str +=  'unhandled|';
        }
  
        str += item.quantity + '|'
        str += item.valid_til  + '|' 
        if (item.simple_posology == undefined) {
            console.log('posology undefined');
          str += '0.00-0.00-0.00-0.00'
        } else {
          formatted =[]
          for (var i = 0; i < item.simple_posology.length; i++) {
            var value = item.simple_posology[i] ? item.simple_posology[i] : 0.00
            str += value.toFixed(2);
            i < item.simple_posology.length-1 ? str += '-' : str += '|'  
          }
        }
        str += item.extended_posology ? item.extended_posology + '|'              : '|'
        str += item.nr_repetitions ? item.nr_repetitions + '|'               : '|'
        str += item.may_be_substituted ? '0|' : '1|'
        s += str;
      }
      s = s.substring(0, s.length -1);
      return s + ';' + this.checksum(s);
    }

    this.checksum = function(string) {
      var sum = 0;
      for (var i = 0; i < string.length; i++) {
        var j = string.charCodeAt(i);
        sum += j;
        // console.log( "checksum i: "+ i + " char <" +string[i] + " -> " + j);
      }
      return sum;
    };    

    this.add_item = function(item2add) {
      this.items.push(item2add);
    };
    
  }
//-";|2014236||1||0.00-0.00-0.00-0.00|||1|||SPEZIALVERBAND|1|20131214|0.00-0.00-0.00-0.00||40|0|7680456740106|||1||1.00-1.00-1.00-0.00|zu Beginn der Mahlzeiten mit mindestens einem halben Glas Wasser einzunehmen||0;27834"
//+";|2014236||1||0.00|0.00|0.00|0.00|||1|||SPEZIALVERBAND|1|20131214|0.00|0.00|0.00|0.00||40|0|7680456740106|||1||1.00|1.00|1.00|0.00|zu Beginn der Mahlzeiten mit mindestens einem halben Glas Wasser einzunehmen||0;28545"

//    QR_TIME_FORMAT          = '%Y%m%d'
//    LOCAL_TIME_FORMAT       = '%d.%m.%Y'
    
}